### PR TITLE
JDK-8317636: Improve heap walking API tests to verify correctness of field indexes

### PR DESCRIPTION
**Pull request by alexmenkov**
The fix adds new test for FollowReferences JVMTI function to verify correctness of reported field indexes.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317636](https://bugs.openjdk.org/browse/JDK-8317636): Improve heap walking API tests to verify correctness of field indexes (**Enhancement** - P4)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to [c5e7b787](https://git.openjdk.org/jdk/pull/17580/files/c5e7b787a5dbfdf94cdeeba66d041486c8d37249)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [a0fffef3](https://git.openjdk.org/jdk/pull/17580/files/a0fffef3f4d5b2ab6d32d4a4c75705949179b326)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17580/head:pull/17580` \
`$ git checkout pull/17580`

Update a local copy of the PR: \
`$ git checkout pull/17580` \
`$ git pull https://git.openjdk.org/jdk.git pull/17580/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17580`

View PR using the GUI difftool: \
`$ git pr show -t 17580`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17580.diff">https://git.openjdk.org/jdk/pull/17580.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17580#issuecomment-1911146988)

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+        fatal(env, "unexpected field type");
+    }
+
+    deallocate(env, name);
+    deallocate(env, sig);
+}
+
+
+/*
+Per jvmtiHeapReferenceInfoField spec:
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:93]: I think you should also include mention that this is for JVMTI_HEAP_REFERENCE_STATIC_FIELD.

Indentation should be fixed to match our C++ comment style, and the "bullets" from the spec should be included to make it clearer that these are two lists of steps.

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+    where n is the count of the fields in all the interfaces implemented by C.
+    Note that C implements all interfaces directly implemented by its superclasses;
+    as well as all superinterfaces of these interfaces.
+If the referrer object is an interface, then the field indices are determined as follows:
+    make a list of the fields directly declared in I.
+    Within this list, put the fields in the order returned by GetClassFields.
+    Assign the fields in this list indices n, n+1, ..., in order,
+    where n is the count of the fields in all the superinterfaces of I.
+
+'Klass' struct contains all required data to calculate field indeces.
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:108]: "indices"

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+    // Initializes fields, field_count.
+    void explore_fields(JNIEnv* env);
+    // Initializes interfaces, interface_count.
+    void explore_interfaces(JNIEnv* env);
+
+    void print() const;
+};
+
+/*
+For each test object 'Object' struct is created and pointer to it set as a tag for jobject.
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:153]:
```suggestion
For each test object, the 'Object' struct is created and a pointer to it is set as the jobject's tag.
```

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+        result += get_max_interface_count(env, interfaces[i]);
+    }
+
+    deallocate(env, interfaces);
+
+    return result;
+}
+
+// Explores all interfaces implemented by 'klass' sorting out duplicates
+// and store them in the 'arr' starting from 'index'.
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:225]:
```suggestion
// Explores all interfaces implemented by 'klass', sorts out duplicates,
// and stores the interfaces in the 'arr' starting from 'index'.
```

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+        if (dup) {
+            continue;
+        }
+
+        // Add the interface.
+        arr[index + count] = Klass::explore(env, interfaces[i]);
+        count++;
+
+        // And explore its superinterfaces.
+        count += fill_interfaces(arr, index + count, env, interfaces[i]);
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:254]: I assume here we are always dealing with test classes that are known not to be deeply nested.

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+
+    fflush(0);
+    return 0;
+}
+
+
+JNIEXPORT void JNICALL
+Java_FieldIndicesTest_prepare(JNIEnv *env, jclass cls, jobject testObj) {
+    Object::explore(env, testObj);
+    fflush(0);
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:492]: stdout?

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+    jlong tag = 0;
+    check_jvmti_status(env,
+        jvmti->GetTag(obj, &tag),
+        "GetTag failed");
+    if (tag != 0) { // already explored
+        return (Object*)tag;
+    }
+
+    jclass obj_klass = env->GetObjectClass(obj);
+
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:348]: Unnecessary newline.

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+    Klass* klass = Klass::explore(env, obj_klass);
+
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:350]: Unnecessary newline.

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 1 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp]: Fixed.

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 2 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp]: Fixed.

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 3 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp]: Fixed.
Updated comment for 'Klass' class the same way

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 4 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp]: Fixed.

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 5 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:255]: Yes, we explore only test classes/interfaces

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 6 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp]: fflush(0) flashes all opened streams (including stout and stderr).
It's to handle the case when some shared test routines log to stderr.
But flash() expects `FILE *`, so I updated all calls to use `nullptr`

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 7 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp]: removed.

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 8 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp]: removed

**Review by plummercj — APPROVED** [review]

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+                                rootObject,                 // initial_object
+                                &heapCallbacks,
+                                nullptr),
+        "FollowReferences failed");
+    fflush(nullptr);
+}
+
+JNIEXPORT jboolean JNICALL
+Java_FieldIndicesTest_testFailed(JNIEnv *env, jclass cls) {
+    return test_failed ? JNI_TRUE : JNI_FALSE;
```
Comment (sspitsyn) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:512]: The indent for native files has to be 2, not 4. Even though there are still some tests with wrong indents I'd suggest for new files to follow this rule.

**Review by sspitsyn — APPROVED** [review]
Nice  test, thank you for developing it!
Looks good.

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 9 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp]: Fixed. Going to integrate the change after sanity testing

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+  }
+  // Replace '/' with '.'.
+  for (char* pos = str; (pos = strchr(pos, '/')) != nullptr; ) {
+    *pos = '.';
+  }
+}
+
+static bool is_static_field(JNIEnv* env, jclass klass, jfieldID fid) {
+  enum {
+   ACC_STATIC        = 0x0008,
```
Comment (sspitsyn) [inline test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp:64]: Nit: Indent is 1 instead of 2.

**Review by sspitsyn — APPROVED** [review]

File: test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp (hunk 10 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp]: fixed.

**Pull request closed without merging**